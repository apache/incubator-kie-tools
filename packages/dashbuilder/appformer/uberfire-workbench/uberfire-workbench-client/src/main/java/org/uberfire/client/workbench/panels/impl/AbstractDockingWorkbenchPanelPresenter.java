/*
 * Copyright 2015 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.uberfire.client.workbench.panels.impl;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import javax.inject.Inject;

import org.uberfire.client.mvp.PerspectiveManager;
import org.uberfire.client.workbench.PanelManager;
import org.uberfire.client.workbench.panels.DockingWorkbenchPanelPresenter;
import org.uberfire.client.workbench.panels.DockingWorkbenchPanelView;
import org.uberfire.client.workbench.panels.WorkbenchPanelPresenter;
import org.uberfire.client.workbench.panels.WorkbenchPanelView;
import org.uberfire.workbench.model.CompassPosition;
import org.uberfire.workbench.model.PanelDefinition;
import org.uberfire.workbench.model.PartDefinition;
import org.uberfire.workbench.model.Position;

import static org.uberfire.client.util.Layouts.heightOrDefault;
import static org.uberfire.client.util.Layouts.widthOrDefault;
import static org.uberfire.client.util.Layouts.widthOrHeight;

/**
 * Implements the behaviour for panel presenters that support adding child panels in {@link CompassPosition} positions.
 */
public abstract class AbstractDockingWorkbenchPanelPresenter<P extends AbstractWorkbenchPanelPresenter<P>>
        extends AbstractWorkbenchPanelPresenter<P> implements DockingWorkbenchPanelPresenter {

    @Inject
    private PanelManager panelManager;

    public AbstractDockingWorkbenchPanelPresenter(WorkbenchPanelView<P> view,
                                                  PerspectiveManager perspectiveManager) {
        super(view,
              perspectiveManager);
    }

    /**
     * Forwards requests to existing child panels in case there is already a child panel in the requested position.
     * Otherwise behaves exactly like the superclass.
     */
    @Override
    public void addPanel(WorkbenchPanelPresenter newChild,
                         Position position) {
        if (getParent() instanceof DockingWorkbenchPanelPresenter) {
            DockingWorkbenchPanelPresenter dockingParent = (DockingWorkbenchPanelPresenter) getParent();
            if (dockingParent.getPanels().get(position) == this) {
                dockingParent.setChildSize(this,
                                           widthOrDefault(newChild.getDefinition()) + widthOrDefault(getDefinition()),
                                           heightOrDefault(newChild.getDefinition()) + heightOrDefault(getDefinition()));
            }
        }
        WorkbenchPanelPresenter existingChild = getPanels().get(position);
        if (existingChild != null && newChild instanceof AbstractDockingWorkbenchPanelPresenter) {
            int existingChildSize = widthOrHeight((CompassPosition) position,
                                                  existingChild.getDefinition());
            int newChildSize = widthOrHeight((CompassPosition) position,
                                             newChild.getDefinition());

            removePanel(existingChild);
            super.addPanel(newChild,
                           position);
            newChild.addPanel(existingChild,
                              position);

            getPanelView().setChildSize(newChild.getPanelView(),
                                        newChildSize + existingChildSize);
        } else {
            super.addPanel(newChild,
                           position);
        }
    }

    /**
     * Checks for existing child panels of the panel to be removed, and reparents them to this panel in the position of
     * the child panel. Once the child panels are safely out of the way, the actual panel removal is done by a call to
     * super.removePanel().
     */
    @Override
    public boolean removePanel(WorkbenchPanelPresenter child) {
        if (child instanceof AbstractDockingWorkbenchPanelPresenter) {

            Position removedPosition = positionOf(child);
            if (removedPosition == null) {
                return false;
            }

            List<AbstractDockingWorkbenchPanelPresenter<?>> rescuedOrphans = new ArrayList<AbstractDockingWorkbenchPanelPresenter<?>>();
            AbstractDockingWorkbenchPanelPresenter<?> dockingChild = (AbstractDockingWorkbenchPanelPresenter<?>) child;
            for (Map.Entry<Position, WorkbenchPanelPresenter> entry : dockingChild.getPanels().entrySet()) {
                dockingChild.removeWithoutOrphanRescue(entry.getValue());
                rescuedOrphans.add((AbstractDockingWorkbenchPanelPresenter<?>) entry.getValue());
                // TODO multiple off-axis orphans need special treatment
                // for example: if the NORTH panel has EAST and WEST children, we need to take the EAST one as a direct child
                // to the NORTH of us, and reparent the other to the WEST of that one (or vice-versa)
                // on the other hand, if there was only one EAST or WEST child, we can just stick it in as our new NORTH child
            }
            super.removePanel(dockingChild);
            for (AbstractDockingWorkbenchPanelPresenter<?> rescued : rescuedOrphans) {
                addPanel(rescued,
                         removedPosition);
            }

            return true;
        } else {
            return super.removePanel(child);
        }
    }

    /**
     * Removes the given child panel without modifying child attachments at all. This is used d
     * @param child
     */
    private boolean removeWithoutOrphanRescue(WorkbenchPanelPresenter child) {
        return super.removePanel(child);
    }

    @Override
    public boolean setChildSize(WorkbenchPanelPresenter child,
                                Integer pixelWidth,
                                Integer pixelHeight) {
        for (Map.Entry<Position, WorkbenchPanelPresenter> e : getPanels().entrySet()) {
            if (e.getValue() == child) {
                int size;
                // Position instance could come from a different script so we convert
                // it to the local type here first in order for == to work.
                CompassPosition cp = CompassPosition.valueOf("" + e.getKey());
                if (cp == CompassPosition.NORTH || cp == CompassPosition.SOUTH) {
                    if (pixelHeight == null) {
                        return false;
                    }
                    size = pixelHeight + nestedPanelHeights(child);
                } else if (cp == CompassPosition.EAST || cp == CompassPosition.WEST) {
                    if (pixelWidth == null) {
                        return false;
                    }
                    size = pixelWidth + nestedPanelWidths(child);
                } else {
                    throw new AssertionError("Unexpected child position: " + e.getKey());
                }

                getPanelView().setChildSize(child.getPanelView(),
                                            size);
                return true;
            }
        }
        return false;
    }

    private int nestedPanelHeights(WorkbenchPanelPresenter child) {
        int totalHeight = 0;
        WorkbenchPanelPresenter northChild = child.getPanels().get(CompassPosition.NORTH);
        if (northChild != null) {
            totalHeight += northChild.getDefinition().getHeight();
            totalHeight += nestedPanelHeights(northChild);
        }
        WorkbenchPanelPresenter southChild = child.getPanels().get(CompassPosition.SOUTH);
        if (southChild != null) {
            totalHeight += southChild.getDefinition().getHeight();
            totalHeight += nestedPanelHeights(southChild);
        }
        return totalHeight;
    }

    private int nestedPanelWidths(WorkbenchPanelPresenter child) {
        int totalWidth = 0;
        WorkbenchPanelPresenter westChild = child.getPanels().get(CompassPosition.WEST);
        if (westChild != null) {
            totalWidth += westChild.getDefinition().getWidth();
            totalWidth += nestedPanelWidths(westChild);
        }
        WorkbenchPanelPresenter eastChild = child.getPanels().get(CompassPosition.EAST);
        if (eastChild != null) {
            totalWidth += eastChild.getDefinition().getWidth();
            totalWidth += nestedPanelWidths(eastChild);
        }
        return totalWidth;
    }

    @Override
    public boolean removePart(PartDefinition part) {
        if (super.removePart(part)) {
            final PanelDefinition panelDef = getDefinition();

            // if we are not the root and we have become empty, we remove ourselves from the panel hierarchy,
            // preserving all child panels
            if (panelDef.getParts().isEmpty() && getParent() != null) {
                panelManager.removeWorkbenchPanel(this.getDefinition());
            }
            return true;
        }
        return false;
    }

    @Override
    public DockingWorkbenchPanelView<P> getPanelView() {
        return (DockingWorkbenchPanelView<P>) super.getPanelView();
    }
}
