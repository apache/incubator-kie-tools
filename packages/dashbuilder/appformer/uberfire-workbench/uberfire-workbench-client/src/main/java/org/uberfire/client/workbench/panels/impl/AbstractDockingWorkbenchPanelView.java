/*
 * Copyright 2015 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.uberfire.client.workbench.panels.impl;

import java.util.IdentityHashMap;
import java.util.Map;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import javax.inject.Inject;

import com.google.gwt.core.client.Scheduler;
import com.google.gwt.core.client.Scheduler.ScheduledCommand;
import com.google.gwt.user.client.ui.DockLayoutPanel.Direction;
import com.google.gwt.user.client.ui.HasWidgets;
import com.google.gwt.user.client.ui.RequiresResize;
import com.google.gwt.user.client.ui.SimpleLayoutPanel;
import com.google.gwt.user.client.ui.Widget;
import org.uberfire.client.workbench.BeanFactory;
import org.uberfire.client.workbench.panels.DockingWorkbenchPanelView;
import org.uberfire.client.workbench.panels.WorkbenchPanelPresenter;
import org.uberfire.client.workbench.panels.WorkbenchPanelView;
import org.uberfire.client.workbench.widgets.dnd.WorkbenchDragAndDropManager;
import org.uberfire.client.workbench.widgets.listbar.ResizeFlowPanel;
import org.uberfire.client.workbench.widgets.split.WorkbenchSplitLayoutPanel;
import org.uberfire.workbench.model.CompassPosition;
import org.uberfire.workbench.model.PanelDefinition;
import org.uberfire.workbench.model.Position;

import static org.uberfire.client.util.Layouts.setToFillParent;
import static org.uberfire.client.util.Layouts.widthOrHeight;
import static org.kie.soup.commons.validation.PortablePreconditions.checkNotNull;
import static org.uberfire.plugin.PluginUtil.toInteger;

/**
 * Implements the view behaviour required by all docking panel views: adding and removing child panels in the NORTH,
 * SOUTH, EAST, and WEST compass positions.
 * <p>
 * <h2>Information for subclassers</h2>
 * The top-level widget of an {@link AbstractDockingWorkbenchPanelView} is always {@link #topLevelWidget}, a container
 * for child panels, even if this panel doesn't currently have any child panels. This is done so child panels can be
 * inserted and removed without making any assumptions about the parent panel this view is located in (if any!).
 * <p>
 * This means you must always put your part view UI into the widget returned from {@link #getPartViewContainer()}. The
 * <i>contents</i> of this container will never be inspected or modified, but the container itself will be reparented as
 * necessary to accommodate child panels being inserted and removed around it. Put another way, <b>do not insert your
 * part view UI directly into the top-level widget of this view! It will get wiped out!</b>
 * <p>
 * This also means you must not call {@link #initWidget(Widget)}. That will always be done by this superclass.
 *
 * @param <P> the presenter type this view binds to
 */
public abstract class AbstractDockingWorkbenchPanelView<P extends WorkbenchPanelPresenter>
        extends AbstractWorkbenchPanelView<P> implements DockingWorkbenchPanelView<P> {

    private final IdentityHashMap<WorkbenchPanelView<?>, WorkbenchSplitLayoutPanel> viewSplitters = new IdentityHashMap<WorkbenchPanelView<?>, WorkbenchSplitLayoutPanel>();

    @Inject
    protected WorkbenchDragAndDropManager dndManager;

    @Inject
    protected BeanFactory factory;

    /**
     * The topmost widget (closest to DOM root) that this panel view manages. Contains either partViewContainer itself
     * (when there are no child panels) or a splitter (when there is at least one child panel).
     */
    private SimpleLayoutPanel topLevelWidget = new SimpleLayoutPanel();

    @Inject
    private ResizeFlowPanel partViewContainer;

    private static CompassPosition toPosition(Direction d) {
        if (d == null) {
            return null;
        }
        switch (d) {
            case NORTH:
                return CompassPosition.NORTH;
            case SOUTH:
                return CompassPosition.SOUTH;
            case EAST:
            case LINE_START:
                return CompassPosition.WEST;
            case WEST:
            case LINE_END:
                return CompassPosition.EAST;
            default:
                throw new IllegalArgumentException("Unknown direction: " + d);
        }
    }

    /**
     * Retrieves the application-requested initial size for a child panel, or calculates a good default based on the
     * available space.
     *
     * @param position   the position the panel will be added within its parent.
     * @param definition the new panel's definition.
     * @param parent     the widget whose space will be used up by the insertion of the new panel.
     */
    static int initialWidthOrHeight(CompassPosition position,
                                    PanelDefinition definition,
                                    Widget parent) {
        Integer requestedSize;
        int availableSize;
        switch (position) {
            case NORTH:
            case SOUTH:
                requestedSize = definition.getHeight();
                availableSize = parent.getOffsetHeight();
                break;
            case EAST:
            case WEST:
                requestedSize = definition.getWidth();
                availableSize = parent.getOffsetWidth();
                break;
            default:
                throw new IllegalArgumentException("Position " + position + " has no horizontal or vertial aspect.");
        }
        if (requestedSize == null || requestedSize <= 0) {
            return availableSize / 2;
        } else {
            return requestedSize;
        }
    }

    static Integer minWidthOrHeight(CompassPosition position,
                                    PanelDefinition definition) {
        switch (position) {
            case NORTH:
            case SOUTH:
                return toInteger(definition.getMinHeightAsInt());
            case EAST:
            case WEST:
                return toInteger(definition.getMinWidthAsInt());
            default:
                throw new IllegalArgumentException("Position " + position + " has no horizontal or vertial aspect.");
        }
    }

    private static int getWidthOrHeight(CompassPosition position,
                                        Widget w) {
        switch (position) {
            case NORTH:
            case SOUTH:
                return w.getOffsetHeight();
            case EAST:
            case WEST:
                return w.getOffsetWidth();
            default:
                throw new IllegalArgumentException("Position " + position + " has no horizontal or vertial aspect.");
        }
    }

    private static void scheduleResize(final RequiresResize widget) {
        Scheduler.get().scheduleDeferred(new ScheduledCommand() {
            @Override
            public void execute() {
                widget.onResize();
            }
        });
    }

    @PostConstruct
    void setupDockingPanel() {
        initWidget(topLevelWidget);
        topLevelWidget.add(partViewContainer);
        setToFillParent(topLevelWidget);
        setToFillParent(partViewContainer);
        if (getPartDropRegion() != null) {
            dndManager.registerDropController(this,
                                              factory.newDropController(this));
        }
    }

    @PreDestroy
    private void tearDownDockingPanel() {
        if (getPartDropRegion() != null) {
            dndManager.unregisterDropController(this);
        }
    }

    /**
     * Overridden to ensure subclasses don't return the partViewContainer by mistake (this would interfere with nested
     * docking panels).
     */
    // override also helps with unit tests: under GWTMockito, super.getWidget() returns a new mock every time
    @Override
    public final Widget getWidget() {
        return topLevelWidget;
    }

    /**
     * Returns the panel that subclasses should put the part view UI into.
     */
    protected ResizeFlowPanel getPartViewContainer() {
        return partViewContainer;
    }

    /**
     * Returns the partViewContainer, which appears to be the "real" on-screen boundary of this widget.
     */
    @Override
    public Widget getPartDropRegion() {
        return getPartViewContainer();
    }

    /**
     * Overridden to attach the ID to the part container rather than the top-level widget, which may contain sub-panels
     * and be larger and further up the DOM tree than desired.
     */
    @Override
    public void setElementId(String elementId) {
        if (elementId == null) {
            getPartViewContainer().getElement().removeAttribute("id");
        } else {
            getPartViewContainer().getElement().setAttribute("id",
                                                             elementId);
        }
    }

    @Override
    public void addPanel(final PanelDefinition childPanelDef,
                         final WorkbenchPanelView<?> childPanelView,
                         final Position childPosition) {

        checkNotNull("childPanelView",
                     childPanelView);
        CompassPosition position = (CompassPosition) checkNotNull("childPosition",
                                                                  childPosition);

        if (viewSplitters.get(position) != null) {
            throw new IllegalStateException("This panel already has a " + position + " child");
        }

        final WorkbenchSplitLayoutPanel splitPanel = new WorkbenchSplitLayoutPanel();
        splitPanel.add(childPanelView.asWidget(),
                       position,
                       widthOrHeight(position,
                                     childPanelDef));

        // now reparent our existing part container into the split panel's resizable area
        Widget partContainerParent = partViewContainer.getParent();
        splitPanel.add(partViewContainer);
        ((HasWidgets) partContainerParent).add(splitPanel); // this is either a WorkbenchSplitLayoutPanel or topLevelWidget (a SimpleLayoutPanel)

        Integer childMinSize = minWidthOrHeight(position,
                                                childPanelDef);
        if (childMinSize != null) {
            splitPanel.setWidgetMinSize(childPanelView.asWidget(),
                                        childMinSize);
        }
        Integer myMinSize = minWidthOrHeight(position,
                                             getPresenter().getDefinition());
        if (myMinSize != null) {
            splitPanel.setWidgetMinSize(splitPanel,
                                        myMinSize);
        }

        viewSplitters.put(childPanelView,
                          splitPanel);

        //Adding an additional embedded ScrollPanel can cause scroll-bars to disappear
        //so ensure we set the sizes of the new Panel and it's children after the
        //browser has added the new DIVs to the HTML tree. This does occasionally
        //add slight flicker when adding a new Panel.
        scheduleResize(splitPanel);
    }

    @Override
    public boolean removePanel(WorkbenchPanelView<?> childView) {
        CompassPosition removalPosition = positionOf(childView);
        if (removalPosition == null) {
            System.out.println("  remove failed - no such child view");
            return false;
        }

        WorkbenchSplitLayoutPanel splitter = viewSplitters.remove(childView);
        splitter.remove(childView.asWidget());

        // now search for 'splitter' in all remaining split panels in the map, plus topLevelWidget
        // when found, transfer orphaned children to the same position as splitter was in in its old parent

        Widget orphan = null;
        for (Widget w : splitter) {
            if (orphan != null) {
                System.out.println("  splitter@" + System.identityHashCode(splitter) + " LOSING ORPHAN: " + splitter.getWidgetDirection(w) + " - " + w);
            }
            orphan = w;
        }

        if (topLevelWidget.getWidget() == splitter) {
            if (orphan != null) {
                topLevelWidget.setWidget(orphan);
            }
        } else {
            for (Map.Entry<WorkbenchPanelView<?>, WorkbenchSplitLayoutPanel> ent : viewSplitters.entrySet()) {
                WorkbenchSplitLayoutPanel sp = ent.getValue();
                if (sp.getWidgetIndex(splitter) >= 0) {
                    Direction d = sp.getWidgetDirection(splitter);
                    Double size = sp.getWidgetSize(splitter);
                    sp.remove(splitter);
                    if (orphan != null) {
                        sp.insert(orphan,
                                  d,
                                  size,
                                  null);
                    }
                }
            }
        }

        scheduleResize(partViewContainer);

        return true;
    }

    private CompassPosition positionOf(WorkbenchPanelView<?> childView) {
        final WorkbenchSplitLayoutPanel splitter = viewSplitters.get(childView);
        if (splitter == null) {
            return null;
        }
        Direction widgetDirection = splitter.getWidgetDirection(childView.asWidget());
        if (widgetDirection == null) {
            throw new AssertionError("Found child in splitter map but not in the splitter itself");
        }
        return toPosition(widgetDirection);
    }

    @Override
    public boolean setChildSize(WorkbenchPanelView<?> childPanel,
                                int size) {
        WorkbenchSplitLayoutPanel splitPanel = viewSplitters.get(childPanel);
        if (splitPanel != null) {
            PanelDefinition definition = getPresenter().getDefinition();
            CompassPosition childPosition = positionOf(childPanel);

            Integer childMinSize = minWidthOrHeight(childPosition,
                                                    definition);
            Integer myMinSize = minWidthOrHeight(childPosition,
                                                 getPresenter().getDefinition());
            int mySize = getWidthOrHeight(childPosition,
                                          asWidget());

            if (childMinSize != null) {
                size = Math.max(size,
                                childMinSize);
            }
            if (myMinSize != null) {
                size = Math.min(size,
                                mySize - myMinSize);
            }

            splitPanel.setWidgetSize(childPanel.asWidget(),
                                     size);

            return true;
        }
        return false;
    }

    /**
     * Overridden to maximize the widget returned by {@link #getPartViewContainer()}.
     */
    @Override
    public void maximize() {
        layoutSelection.get().maximize(getPartViewContainer());
    }

    /**
     * Overridden to match {@link #maximize()}.
     */
    @Override
    public void unmaximize() {
        layoutSelection.get().unmaximize(getPartViewContainer());
    }
}
