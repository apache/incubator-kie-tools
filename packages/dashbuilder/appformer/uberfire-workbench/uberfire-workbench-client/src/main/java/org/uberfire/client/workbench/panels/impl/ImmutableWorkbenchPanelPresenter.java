package org.uberfire.client.workbench.panels.impl;

import javax.enterprise.context.Dependent;
import javax.inject.Inject;
import javax.inject.Named;

import org.uberfire.client.mvp.PerspectiveManager;
import org.uberfire.client.mvp.PlaceManager;
import org.uberfire.client.workbench.part.WorkbenchPartPresenter;

/**
 * An undecorated panel that can contain one part at a time and does not support child panels. The part's view fills
 * the entire panel. Does not support changing the part after it is set. Does not support drag-and-drop rearrangement of
 * parts.
 */
@Dependent
public class ImmutableWorkbenchPanelPresenter extends AbstractWorkbenchPanelPresenter<ImmutableWorkbenchPanelPresenter> {

    private PlaceManager placeManager;

    @Inject
    public ImmutableWorkbenchPanelPresenter(@Named("ImmutableWorkbenchPanelView") final ImmutableWorkbenchPanelView view,
                                            final PerspectiveManager perspectiveManager,
                                            final PlaceManager placeManager) {
        super(view, perspectiveManager);
        this.placeManager = placeManager;
    }

    @Override
    protected ImmutableWorkbenchPanelPresenter asPresenterType() {
        return this;
    }

    /**
     * Returns null (static panels don't support child panels).
     */
    @Override
    public String getDefaultChildType() {
        return null;
    }

    @Override
    public void addPart(final WorkbenchPartPresenter part) {
        this.addPart(part, null);
    }

    @Override
    public void addPart(final WorkbenchPartPresenter part,
                        final String contextId) {
        if (createSinglePartPanelHelper().hasNoParts()) {
            super.addPart(part, contextId);
        }
    }

    SinglePartPanelHelper createSinglePartPanelHelper() {
        return new SinglePartPanelHelper(getPanelView().getParts(), placeManager);
    }
}
