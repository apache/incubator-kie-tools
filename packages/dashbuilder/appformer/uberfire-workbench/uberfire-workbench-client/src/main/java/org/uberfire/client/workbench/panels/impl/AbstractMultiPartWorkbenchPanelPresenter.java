/*
 * Copyright 2015 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.uberfire.client.workbench.panels.impl;

import java.util.HashMap;
import java.util.Map;

import org.uberfire.client.mvp.ActivityManager;
import org.uberfire.client.mvp.ContextActivity;
import org.uberfire.client.mvp.PerspectiveManager;
import org.uberfire.client.workbench.panels.WorkbenchPanelView;
import org.uberfire.client.workbench.part.WorkbenchPartPresenter;
import org.uberfire.mvp.impl.DefaultPlaceRequest;
import org.uberfire.workbench.model.ContextDefinition;
import org.uberfire.workbench.model.ContextDisplayMode;
import org.uberfire.workbench.model.PanelDefinition;
import org.uberfire.workbench.model.PartDefinition;

import static org.uberfire.workbench.model.ContextDisplayMode.SHOW;

public abstract class AbstractMultiPartWorkbenchPanelPresenter<P extends AbstractMultiPartWorkbenchPanelPresenter<P>>
        extends AbstractDockingWorkbenchPanelPresenter<P> {

    private final Map<PartDefinition, ContextActivity> contextActivities = new HashMap<PartDefinition, ContextActivity>();
    protected ActivityManager activityManager;
    private ContextActivity perspectiveContext = null;
    private ContextActivity panelContext = null;

    protected AbstractMultiPartWorkbenchPanelPresenter(final WorkbenchPanelView<P> view,
                                                       final ActivityManager activityManager,
                                                       final PerspectiveManager perspectiveManager) {
        super(view,
              perspectiveManager);
        this.activityManager = activityManager;
    }

    private void buildPerspectiveContext() {
        final ContextDefinition contextDefinition = perspectiveManager.getLivePerspectiveDefinition().getContextDefinition();
        final ContextDisplayMode contextDisplayMode = perspectiveManager.getLivePerspectiveDefinition().getContextDisplayMode();
        if (contextDefinition != null && contextDisplayMode == SHOW) {
            final ContextActivity activity = activityManager.getActivity(ContextActivity.class,
                                                                         contextDefinition.getPlace());
            if (activity != null) {
                perspectiveContext = activity;
            }
        }
    }

    @Override
    public void setDefinition(final PanelDefinition definition) {
        super.setDefinition(definition);

        final ContextDisplayMode perspectiveContextDisplayMode = perspectiveManager.getLivePerspectiveDefinition().getContextDisplayMode();

        if (definition.getContextDefinition() != null
                && perspectiveContextDisplayMode == SHOW
                && definition.getContextDisplayMode() == SHOW) {
            final ContextActivity activity = activityManager.getActivity(ContextActivity.class,
                                                                         definition.getContextDefinition().getPlace());
            if (activity != null) {
                panelContext = activity;
            }
        }
        buildPerspectiveContext();
    }

    @Override
    public void addPart(final WorkbenchPartPresenter part,
                        final String contextId) {
        super.addPart(part,
                      contextId);
        final ContextDisplayMode perspectiveContextDisplayMode = perspectiveManager.getLivePerspectiveDefinition().getContextDisplayMode();
        if (perspectiveContextDisplayMode == SHOW
                && getDefinition().getContextDisplayMode() == SHOW
                && part.getDefinition().getContextDisplayMode() == SHOW) {
            ContextActivity contextActivity = null;
            if (contextId != null) {
                contextActivity = activityManager.getActivity(ContextActivity.class,
                                                              new DefaultPlaceRequest(contextId));
            } else if (part.getDefinition().getContextDefinition() != null) {
                contextActivity = activityManager.getActivity(ContextActivity.class,
                                                              part.getDefinition().getContextDefinition().getPlace());
            } else if (part.getContextId() != null) {
                contextActivity = activityManager.getActivity(ContextActivity.class,
                                                              new DefaultPlaceRequest(part.getContextId()));
            }
            if (contextActivity != null) {
                contextActivities.put(part.getDefinition(),
                                      contextActivity);
            }
        }
    }

    @Override
    public boolean removePart(final PartDefinition part) {
        boolean removed = super.removePart(part);
        contextActivities.remove(part);
        return removed;
    }

    public ContextActivity resolveContext(final PartDefinition part) {
        ContextActivity result = perspectiveContext;
        if (panelContext != null) {
            result = panelContext;
        }
        if (contextActivities.containsKey(part)) {
            result = contextActivities.get(part);
        }
        return result;
    }
}
