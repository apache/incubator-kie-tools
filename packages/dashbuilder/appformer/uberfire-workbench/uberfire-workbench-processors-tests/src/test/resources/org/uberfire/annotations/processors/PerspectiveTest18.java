package org.uberfire.annotations.processors;

import javax.annotation.PostConstruct;
import javax.enterprise.context.ApplicationScoped;

import com.google.gwt.user.client.ui.FlowPanel;

import org.uberfire.client.annotations.WorkbenchPanel;
import org.uberfire.client.annotations.WorkbenchPerspective;
import org.uberfire.client.workbench.panels.impl.MultiListWorkbenchPanelPresenter;
import org.jboss.errai.common.client.dom.Div;
import org.jboss.errai.ui.client.local.api.IsElement;
import javax.inject.Inject;

@ApplicationScoped
@WorkbenchPerspective(
        identifier = "HomePerspective",
        isDefault = true)
public class PerspectiveTest18 implements IsElement {

    @Inject
    @WorkbenchPanel(panelType = MultiListWorkbenchPanelPresenter.class, isDefault = true, parts = "TesteScreen")
    Div teste1;

    @Inject
    @WorkbenchPanel(parts = { "HelloWorldScreen1", "HelloWorldScreen2" })
    Div teste2;

    @PostConstruct
    public void setup() {

    }


}