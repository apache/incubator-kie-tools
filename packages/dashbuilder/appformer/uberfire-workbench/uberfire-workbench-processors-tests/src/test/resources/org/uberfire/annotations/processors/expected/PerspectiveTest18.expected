/*
 * Copyright 2012 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.uberfire.annotations.processors;

import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.function.Consumer;
import javax.annotation.Generated;
import javax.enterprise.context.Dependent;
import javax.inject.Inject;

import javax.inject.Named;
import org.uberfire.workbench.model.PerspectiveDefinition;
import org.uberfire.workbench.model.impl.PerspectiveDefinitionImpl;
import java.util.HashMap;
import java.util.Map;
import org.uberfire.client.annotations.Perspective;
import org.jboss.errai.common.client.dom.HTMLElement;
import org.jboss.errai.common.client.dom.DOMUtil;
import org.uberfire.client.mvp.IsTemplatePerspective;
import org.uberfire.client.mvp.AbstractWorkbenchPerspectiveActivity;
import org.uberfire.client.mvp.PlaceManager;

import org.uberfire.mvp.PlaceRequest;

import com.google.gwt.user.client.ui.HasWidgets;
import com.google.gwt.user.client.ui.IsWidget;
import org.uberfire.client.mvp.TemplatedActivity;
import org.uberfire.client.workbench.panels.impl.TemplatedWorkbenchPanelPresenter;
import org.uberfire.mvp.impl.DefaultPlaceRequest;
import org.uberfire.workbench.model.NamedPosition;
import org.uberfire.workbench.model.PanelDefinition;
import org.uberfire.workbench.model.Position;
import org.uberfire.workbench.model.impl.PanelDefinitionImpl;
import org.uberfire.workbench.model.impl.PartDefinitionImpl;

@Dependent
@Generated("org.uberfire.annotations.processors.WorkbenchPerspectiveProcessor")
@Named("HomePerspective")
@IsTemplatePerspective
/*
 * WARNING! This class is generated. Do not modify.
 */
public class PerspectiveTest18Activity extends AbstractWorkbenchPerspectiveActivity implements TemplatedActivity {

    private static final String UF_PERSPECTIVE_COMPONENT = "uf-perspective-component";

    @Inject
    private PerspectiveTest18 realPresenter;

    @Inject
    //Constructor injection for testing
    public PerspectiveTest18Activity(final PlaceManager placeManager) {
        super( placeManager );
    }

    @Override
    public String getIdentifier() {
        return "HomePerspective";
    }

    @Override
    public boolean isDefault() {
        return true;
    }

    @Override
    public HTMLElement getRootElement() {
        return realPresenter.getElement();
    }

    @Override
    public HTMLElement resolvePosition( NamedPosition position ) {
        final String fieldName = position.getName();
        if ( fieldName.equals( "teste1" ) ) {
            DOMUtil.addCSSClass( realPresenter.teste1, UF_PERSPECTIVE_COMPONENT );
            return realPresenter.teste1;
        }
        if ( fieldName.equals( "teste2" ) ) {
            DOMUtil.addCSSClass( realPresenter.teste2, UF_PERSPECTIVE_COMPONENT );
            return realPresenter.teste2;
        }
        return null;
    }

    @Override
    public PerspectiveDefinition getDefaultPerspectiveLayout() {
        final PerspectiveDefinition p = new PerspectiveDefinitionImpl( TemplatedWorkbenchPanelPresenter.class.getName() );
        p.setName( realPresenter.getClass().getName() );

        PanelDefinition panelDefinition = new PanelDefinitionImpl( "org.uberfire.client.workbench.panels.impl.MultiListWorkbenchPanelPresenter" );
        panelDefinition.addPart(
            new PartDefinitionImpl(new DefaultPlaceRequest( "TesteScreen" ) ) );
        p.getRoot().appendChild( new NamedPosition( "teste1" ), panelDefinition );

        PanelDefinition panelDefinition0 = new PanelDefinitionImpl( "PARENT_CHOOSES_TYPE" );
        panelDefinition0.addPart(
                new PartDefinitionImpl( new DefaultPlaceRequest( "HelloWorldScreen1" ) ) );
        panelDefinition0.addPart(
                new PartDefinitionImpl( new DefaultPlaceRequest( "HelloWorldScreen2" ) ) );
        p.getRoot().appendChild( new NamedPosition( "teste2" ), panelDefinition0 );
        return p;
    }
}